### PR TITLE
Add support for missing video quality related APIs and events

### PR DESCRIPTION
**Pull request by 123mpozzi**
## Description

Add support for missing video quality related APIs and events

## Changes

- Add getter for `videoQuality` Player API
- Add getter for `availableQualities` Player API
- Add `VideoDownloadQualityChanged` event
- iOS: Fix `VideoPlaybackQualityChanged` event

## Testing
Apply the following patch:
[videoQuality.patch](https://github.com/bitmovin/bitmovin-player-react-native/files/13283621/videoQuality.patch)

Check for:
- `getVideoQuality` updates correctly. Current video quality is logged on _onPlay_ triggers
- `getAvailableVideoQualities` updates correctly. Current available qualities are logged on _onPlay_ triggers
- `VideoDownloadQualityChanged` event is emitted correctly
- iOS: `VideoPlaybackQualityChanged` event is emitted correctly

## Checklist
- [x] 🗒 `CHANGELOG` entry

### CODE REVIEW

File: src/player.ts [hunk 1]
```diff
@@ -405,4 +405,18 @@ export class Player extends NativeInstance<PlayerConfig> {
   castStop = () => {
     PlayerModule.castStop(this.nativeId);
   };
+
+  /**
```
Comment (rolandkakonyi) [inline src/player.ts:409]: We don't have a description here. Please see the other methods that have a description and the return value documented.

File: src/player.ts [hunk 2]
```diff
@@ -405,4 +405,18 @@ export class Player extends NativeInstance<PlayerConfig> {
...
+   * @returns The currently selected video quality.
+   */
+  getVideoQuality = async (): Promise<number> => {
+    return PlayerModule.getVideoQuality(this.nativeId);
+  };
+
+  /**
```
Comment (rolandkakonyi) [inline src/player.ts:416]: Same here

File: src/player.ts [hunk 3]
```diff
@@ -405,4 +405,18 @@ export class Player extends NativeInstance<PlayerConfig> {
   castStop = () => {
     PlayerModule.castStop(this.nativeId);
   };
+
+  /**
+   * @returns The currently selected video quality.
+   */
+  getVideoQuality = async (): Promise<number> => {
```
Comment (rolandkakonyi) [inline src/player.ts:412]: This promise has wrong type, please use `VideoQuality`.
```suggestion
  getVideoQuality = async (): Promise<VideoQuality> => {
```

File: ios/PlayerModule.swift [hunk 4]
```diff
@@ -686,4 +686,38 @@ public class PlayerModule: NSObject, RCTBridgeModule { // swiftlint:disable:this
...
+     - Parameter rejecter: JS promise rejecter.
+     */
+    @objc(getVideoQuality:resolver:rejecter:)
+    func getVideoQuality(
+        _ nativeId: NativeId,
+        resolver resolve: @escaping RCTPromiseResolveBlock,
+        rejecter reject: @escaping RCTPromiseRejectBlock
+    ) {
+        bridge.uiManager.addUIBlock { [weak self] _, _ in
+            resolve(self?.players[nativeId]?.videoQuality)
```
Comment (rolandkakonyi) [inline ios/PlayerModule.swift:703]: You need to convert the `VideoQuality` type to JSON before resolving.
Please use `RCTConvert.toJson(videoQuality:)` for this.

File: ios/PlayerModule.swift [hunk 5]
```diff
@@ -686,4 +686,38 @@ public class PlayerModule: NSObject, RCTBridgeModule { // swiftlint:disable:this
...
+     - Parameter rejecter: JS promise rejecter.
+     */
+    @objc(getAvailableVideoQualities:resolver:rejecter:)
+    func getAvailableVideoQualities(
+        _ nativeId: NativeId,
+        resolver resolve: @escaping RCTPromiseResolveBlock,
+        rejecter reject: @escaping RCTPromiseRejectBlock
+    ) {
+        bridge.uiManager.addUIBlock { [weak self] _, _ in
+            resolve(self?.players[nativeId]?.availableVideoQualities)
```
Comment (rolandkakonyi) [inline ios/PlayerModule.swift:720]: You need to convert the `VideoQuality` type to JSON before resolving.
Please use `RCTConvert.toJson(videoQuality:)` for this.

See how it is done with an array in `getAvailableAudioTracks`

**Review by rolandkakonyi — CHANGES_REQUESTED** [review]

File: ios/PlayerModule.swift (hunk 5 above)
Comment (123mpozzi) [reply ios/PlayerModule.swift]: Address in 82db9efcb6b79230df50f145700dbefb7dd20d33

File: ios/PlayerModule.swift (hunk 4 above)
Comment (123mpozzi) [reply ios/PlayerModule.swift]: Address in 82db9efcb6b79230df50f145700dbefb7dd20d33

File: src/player.ts (hunk 3 above)
Comment (123mpozzi) [reply src/player.ts]: Address in efb83e696e92f9db67998973f270e787b2da8d2c and 345c7fe21d44c7441167882021f901b390c29512

File: src/player.ts [hunk 6]
```diff
@@ -405,4 +405,18 @@ export class Player extends NativeInstance<PlayerConfig> {
...
+    return PlayerModule.getVideoQuality(this.nativeId);
+  };
+
+  /**
+   * @returns An array containing all available video qualities the player can adapt between.
+   */
+  getAvailableVideoQualities = async (): Promise<number> => {
```
[the remark this answers is not in the document]
Comment (123mpozzi) [reply src/player.ts]: Address in efb83e696e92f9db67998973f270e787b2da8d2c and 345c7fe21d44c7441167882021f901b390c29512

File: android/src/main/java/com/bitmovin/player/reactnative/PlayerModule.kt [hunk 7]
```diff
@@ -543,6 +543,30 @@ class PlayerModule(private val context: ReactApplicationContext) : ReactContextB
...
 
+    /**
+     * Resolve `nativeId`'s current video quality.
+     * @param nativeId Target player Id.
+     * @param promise JS promise object.
+     */
+    @ReactMethod
+    fun getVideoQuality(nativeId: NativeId, promise: Promise) {
+        uiManager()?.addUIBlock {
+            promise.resolve(players[nativeId]?.source?.selectedVideoQuality)
```
[the remark this answers is not in the document]
Comment (123mpozzi) [reply android/src/main/java/com/bitmovin/player/reactnative/PlayerModule.kt]: Address in 82db9efcb6b79230df50f145700dbefb7dd20d33

File: src/player.ts (hunk 2 above)
Comment (123mpozzi) [reply src/player.ts:418]: Address in cafb79f9ec5399959d523103108a7cf581495c07

File: src/player.ts (hunk 1 above)
Comment (123mpozzi) [reply src/player.ts:410]: Address in cafb79f9ec5399959d523103108a7cf581495c07

**Review by rolandkakonyi — APPROVED** [review]
Nice job!

**Review by matamegger — APPROVED** [review]
👍

**Pull request merged**